modules/scheduler-api/src/main/java/org/opencastproject/scheduler/api/Util.java

```diff
@@ -44,58 +53,89 @@ private Util() {
+  public static List<Period> calculatePeriods(Date startUtc, Date endUtc, long duration, RRule rRule, TimeZone tz) {
+    List<Period> events = new LinkedList<>();
+    TimeZoneRegistry registry = TimeZoneRegistryFactory.getInstance().createRegistry();
+
+    logger.debug("Inbound start of recurrence {} and end of recurrence {}", startUtc, endUtc);
+    DateTime startInTz = new DateTime(startUtc).toDateTime(DateTimeZone.forTimeZone(tz));
+    DateTime endInTz = new DateTime(endUtc).toDateTime(DateTimeZone.forTimeZone(tz));
+
+    net.fortuna.ical4j.model.DateTime periodStartTz = new net.fortuna.ical4j.model.DateTime(startInTz.toDate());
+    net.fortuna.ical4j.model.DateTime periodEndTz = new net.fortuna.ical4j.model.DateTime(endInTz.toDate());
+
+    Calendar endCalendarTz = Calendar.getInstance(tz);
+    endCalendarTz.setTime(periodEndTz);
+
+    Calendar calendarTz = Calendar.getInstance(tz);
+    calendarTz.setTime(periodStartTz);
+
+    calendarTz.set(Calendar.DAY_OF_MONTH, endCalendarTz.get(Calendar.DAY_OF_MONTH));
+    calendarTz.set(Calendar.MONTH, endCalendarTz.get(Calendar.MONTH));
+    calendarTz.set(Calendar.YEAR, endCalendarTz.get(Calendar.YEAR));
+    periodEndTz.setTime(calendarTz.getTime().getTime() + duration);
     duration = duration % (DateTimeConstants.MILLIS_PER_DAY);
 
-    List<Period> events = new LinkedList<>();
+    logger.debug("1-Looking at recurrences for {} to {}, duration {}", periodStartTz.getTime(), periodEndTz.toInstant().toString(), duration);
+    // Have to change the TimeZone to UTC for the rRule.getRecur() to work correctly in a non-global TimeZone
+    periodStartTz.setTimeZone(registry.getTimeZone("UTC"));
+    periodEndTz.setTimeZone(registry.getTimeZone("UTC"));
 
-    TimeZone.setDefault(utc);
-    for (Object date : rRule.getRecur().getDates(periodStart, periodEnd, net.fortuna.ical4j.model.parameter.Value.DATE_TIME)) {
-      Date d = (Date) date;
-      Calendar cDate = Calendar.getInstance(utc);
+    // Special case for first Sunday in a DST change recurrence
+    // Sunday DST day bug: https://github.com/ical4j/ical4j/issues/117
+    boolean firstSundaySpecialCase = false;
+    if (((tz.inDaylightTime(periodStartTz) && !tz.inDaylightTime(periodEndTz))
+            || (!tz.inDaylightTime(periodStartTz) && tz.inDaylightTime(periodEndTz)))) {
+      firstSundaySpecialCase = true;
+    }
 
-      // Adjust for DST, if start of event
-      if (tz.inDaylightTime(periodStart)) { // Event starts in DST
-        if (!tz.inDaylightTime(d)) { // Date not in DST?
-          d.setTime(d.getTime() + tz.getDSTSavings()); // Adjust for Fall back one hour
-        }
-      } else { // Event doesn't start in DST
-        if (tz.inDaylightTime(d)) {
-          d.setTime(d.getTime() - tz.getDSTSavings()); // Adjust for Spring forward one hour
+    for (Object date : rRule.getRecur().getDates(periodStartTz, periodEndTz, net.fortuna.ical4j.model.parameter.Value.DATE_TIME)) {
```

Comment:
Stepping through calculateDaysChange, I think there might be a logic bug in the RRule generation logic.  We have a rule, a start, and an end.  The start and end are exactly 5 days + duration apart, except that the timezone changes mid period, meaning the values are 5 days + duration - 1 hour apart.  In other words, periodStartTz is 0905, periodEndTz is 0*8*10, and the RRule says create a new event at 0900.  See the issue?  The rule is being followed to the letter, rather than what we meant.  So the question is, do we need two RRules?  Or should we do everything in UTC, and convert the RRule's time to UTC as well?

Reply:
@gregorydlogan I think you are onto something, the latest mnode OSGI-compatible ical4j version of the fortuna ical4j seems to resolve some of the Sunday-schedule-in-the-middle-of-the-DST. But, it still has problems when the recurrence starts or ends in the middle of a DST change. The oddities are visible when you change your machine clock to  be on one side of the UTC and then test a recurrence with time zone on the other side of UTC and one of the scheduled dates falls on the DST day.

Reply:
i.e. just when you think you've fixed the issue and then swap your machine time to be on the other side of the UTC, and the test schedule to be on the side, the issue re-manifests.

Reply:
The ical4j code is on github. I suspect is may introduce machine time zone at some point in the recurrence calculation instead of sticking to UTC time.

Reply:
Looking at this more today, the recurrence logic internal to ical4j is totally broken.  For a period of 2016-03-25 - 2016-03-29 (the last item in calculateDaysChange in UtilTests), the recurrence code is generating five candidates:
- 2016-03-2**1**
- 2016-03-24
- 2016-03-25
- 2016-03-26
- 2016-03-27


Note the bolding on the first date.

I'll look further into their code tomorrow I guess.  I'm not sure if we're doing something wrong, but generating a date that far out of the possible range is very strange to me.  That's not an I-have-a-timezone bug...

Reply:
Further work, while my brains dribble out my ears.  The problematic code lives here:

```
getAbsWeekDays:1031, Recur (net.fortuna.ical4j.model)
getDayVariants:996, Recur (net.fortuna.ical4j.model)
getCandidates:818, Recur (net.fortuna.ical4j.model)
getDates:627, Recur (net.fortuna.ical4j.model)
getDates:520, Recur (net.fortuna.ical4j.model)
calculatePeriods:102, Util (org.opencastproject.scheduler.api)
generatePeriods:172, UtilTests (org.opencastproject.scheduler.api)
calculateDaysChange:90, UtilTests (org.opencastproject.scheduler.api)
```

Which lives [here](https://github.com/ical4j/ical4j/blob/develop/src/main/java/net/fortuna/ical4j/model/Recur.java#L1031).  I'm not 100% sure what exactly they're trying to do, but the issue stems from line 1033, where it sets the calendar to the first day of the start date's week.  This is 2018-03-21, which explains where *that* date is coming from.  It does not, however, explain how the rest of the dates are correct.  I'm too tired to continue tonight, but I'm very very sure that a large part of our current set of issues stem from the underlying ical4j library.

Reply:
This is now filed [here](https://github.com/ical4j/ical4j/issues/232) as well

Reply:
Hi @gregorydlogan , using the latest ical4j and ensuring that the inbound recurrence dates are in the TZ that is being scheduled to (not UTC), the recurrence result dates and times are looking good and no fancy adjustments are needed anymore outside of the ical4j code. Test results are now identical when run on a system with local  TZ on either side of UTC. I'm doing some more sanity checks before pushing to the t/MH-12177-TimeZone-Issue-scheduler-dev branch. I've locally rebased that branch to OC develop. But this commit should be applicable to 4x, 5x.